File: bindings/ruby/bench.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require "fiddle"
require "fiddle/import"
require "json"
require "optparse"

options = {
  cases: File.expand_path("../../benchmarks/cases/postgres_sqlite.jsonl", __dir__),
  library: File.expand_path("../../target/release/libsqlgrok.dylib", __dir__),
  iterations: 1000,
  samples: 5,
  warmup: 100
}

OptionParser.new do |opts|
  opts.on("--cases PATH") { |value| options[:cases] = File.expand_path(value) }
  opts.on("--library PATH") { |value| options[:library] = File.expand_path(value) }
  opts.on("--iterations N", Integer) { |value| options[:iterations] = value }
  opts.on("--samples N", Integer) { |value| options[:samples] = value }
  opts.on("--warmup N", Integer) { |value| options[:warmup] = value }
end.parse!

cases = File.readlines(options[:cases], chomp: true)
            .reject { |line| line.strip.empty? || line.lstrip.start_with?("#") }
            .map { |line| JSON.parse(line) }

lib = Fiddle.dlopen(options[:library])
transpile = Fiddle::Function.new(
  lib["sqlgrok_transpile"],
  [Fiddle::TYPE_VOIDP, Fiddle::TYPE_VOIDP, Fiddle::TYPE_VOIDP],
  Fiddle::TYPE_VOIDP
)
free = Fiddle::Function.new(lib["sqlgrok_free"], [Fiddle::TYPE_VOIDP], Fiddle::TYPE_VOID)

checksum = 0
call = lambda do |case_|
  ptr = transpile.call(case_["sql"], case_["read"], case_["write"])
  raise "transpile failed for #{case_['id']}" if ptr.null?

  begin
    output = ptr.to_s
    checksum = (checksum + output.length) & 0xffffffff
  ensure
    free.call(ptr)
  end
end

options[:warmup].times do
  cases.each { |case_| call.call(case_) }
end

def percentile(values, q)
  return 0.0 if values.empty?

  sorted = values.sort
  index = ((q / 100.0) * sorted.length).ceil - 1
  sorted[[[index, 0].max, sorted.length - 1].min]
end

samples = []
measured_checksum = 0
options[:samples].times do
  checksum = 0
  started = Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond)
  options[:iterations].times do
    cases.each { |case_| call.call(case_) }
  end
  elapsed_ns = Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond) - started
  measured_checksum = checksum
  samples << {
    elapsed_ns: elapsed_ns,
    ns_per_op: elapsed_ns.to_f / (cases.length * options[:iterations]),
    checksum: checksum
  }
end
operations = cases.length * options[:iterations]
ns_per_op = samples.map { |sample| sample[:ns_per_op] }

puts JSON.generate(
  binding: "ruby-fiddle",
  checksum: measured_checksum,
  cases: cases.length,
  iterations: options[:iterations],
  samples: options[:samples],
  operations: operations,
  min_ns_per_op: ns_per_op.min,
  mean_ns_per_op: ns_per_op.sum / ns_per_op.length,
  median_ns_per_op: percentile(ns_per_op, 50),
  p95_ns_per_op: percentile(ns_per_op, 95),
  max_ns_per_op: ns_per_op.max,
  per_sample: samples
)

File: bindings/node/bench.js
const fs = require("fs");
const path = require("path");
const koffi = require("koffi");

function readCases(file) {
  return fs
    .readFileSync(file, "utf8")
    .split(/\r?\n/)
    .filter((line) => line.trim() && !line.trimStart().startsWith("#"))
    .map((line) => JSON.parse(line));
}

function parseArgs(argv) {
  const args = {
    cases: path.resolve(__dirname, "../../benchmarks/cases/postgres_sqlite.jsonl"),
    library: path.resolve(__dirname, "../../target/release/libsqlgrok.dylib"),
    iterations: 1000,
    samples: 5,
    warmup: 100,
  };
  for (let i = 2; i < argv.length; i += 1) {
    const arg = argv[i];
    if (arg === "--cases") args.cases = path.resolve(argv[++i]);
    else if (arg === "--library") args.library = path.resolve(argv[++i]);
    else if (arg === "--iterations") args.iterations = Number(argv[++i]);
    else if (arg === "--samples") args.samples = Number(argv[++i]);
    else if (arg === "--warmup") args.warmup = Number(argv[++i]);
    else throw new Error(`unknown argument ${arg}`);
  }
  return args;
}

function percentile(values, q) {
  if (values.length === 0) return 0;
  const sorted = [...values].sort((a, b) => a - b);
  const index = Math.ceil((q / 100) * sorted.length) - 1;
  return sorted[Math.max(0, Math.min(sorted.length - 1, index))];
}

function summarize(values) {
  const sum = values.reduce((acc, value) => acc + value, 0);
  return {
    min_ns_per_op: Math.min(...values),
    mean_ns_per_op: sum / values.length,
    median_ns_per_op: percentile(values, 50),
    p95_ns_per_op: percentile(values, 95),
    max_ns_per_op: Math.max(...values),
  };
}

function main() {
  const args = parseArgs(process.argv);
  const cases = readCases(args.cases);
  const lib = koffi.load(args.library);
  const free = lib.func("sqlgrok_free", "void", ["void *"]);
  const RustString = koffi.disposable("RustString", "str", free);
  const transpile = lib.func("sqlgrok_transpile", RustString, ["str", "str", "str"]);

  let checksum = 0;
  function call(case_) {
    const output = transpile(case_.sql, case_.read, case_.write);
    if (output == null) throw new Error(`transpile failed for ${case_.id}`);
    checksum = (checksum + output.length) >>> 0;
  }

  for (let i = 0; i < args.warmup; i += 1) {
    for (const case_ of cases) call(case_);
  }

  const samples = [];
  let measuredChecksum = 0;
  for (let sample = 0; sample < args.samples; sample += 1) {
    checksum = 0;
    const started = process.hrtime.bigint();
    for (let i = 0; i < args.iterations; i += 1) {
      for (const case_ of cases) call(case_);
    }
    const elapsedNs = Number(process.hrtime.bigint() - started);
    measuredChecksum = checksum;
    samples.push({
      elapsed_ns: elapsedNs,
      ns_per_op: elapsedNs / (cases.length * args.iterations),
      checksum,
    });
  }

  const nsPerOp = samples.map((sample) => sample.ns_per_op);
  console.log(
    JSON.stringify({
      binding: "node-koffi",
      checksum: measuredChecksum,
      cases: cases.length,
      iterations: args.iterations,
      samples: args.samples,
      operations: cases.length * args.iterations,
      ...summarize(nsPerOp),
      per_sample: samples,
    })
  );
}

main();
